src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/dense_set.h"
+
+#include <absl/numeric/bits.h>
+
+#include <cstddef>
+#include <stack>
+#include <type_traits>
+#include <vector>
+
+#include "glog/logging.h"
+
+extern "C" {
+#include "redis/zmalloc.h"
+}
+
+namespace dfly {
+using namespace std;
+
+constexpr size_t kMinSizeShift = 2;
+constexpr size_t kMinSize = 1 << kMinSizeShift;
+constexpr bool kAllowDisplacements = true;
+constexpr bool kFastScan = true;
+
+DenseSet::ChainIterator& DenseSet::ChainIterator::operator++() {
+  if (curr_) {
+    curr_ = curr_->Next();
+  }
+
+  return *this;
+}
+
+DenseSet::ConstChainIterator& DenseSet::ConstChainIterator::operator++() {
+  if (curr_) {
+    curr_ = curr_->Next();
+  }
+
+  return *this;
+}
+
+bool DenseSet::IsEmpty(DenseSet::ChainVectorIterator it) const {
+  return it->IsEmpty();
+}
+
+size_t DenseSet::PushFront(DenseSet::ChainVectorIterator it, void* data) {
+  // if this is an empty list assign the value to the empty placeholder pointer
+  if (it->IsEmpty()) {
+    it->SetObject(data);
+    return ObjectAllocSize(data);
+  }
+
+  // otherwise make a new link and connect it to the front of the list
+  LinkAllocator la(mr());
+  DenseLinkKey* lk = la.allocate(1);
+  la.construct(lk);
+
+  lk->next = *it;
+  it->SetLink(lk);
+  lk->SetObject(data);
+
+  return ObjectAllocSize(data);
+}
+
+void DenseSet::PushFront(DenseSet::ChainVectorIterator it, DenseSet::DensePtr* ptr) {
+  if (it->IsEmpty()) {
+    it->SetObject(ptr->GetObject());
+    if (ptr->IsLink()) {
+      // deallocate the link if it is no longer a link as it is now in an empty list
+      mr()->deallocate(ptr->Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+    }
+  } else if (ptr->IsLink()) {
+    // if the pointer is already a link then no allocation needed
+    *ptr->Next() = *it;
```

Comment:
This is the line that "requires" that ptr will be passed by pointer: you change its contents. I expected to see that the caller uses the result and it's important. But then, when I looked at the caller context at line 266, I saw no one uses ptr afterwards. Then why passing `DensePtr*`?